keps/sig-node/0036-20190118-ForceEviction.md

```diff
@@ -0,0 +1,178 @@
+pods to start. The main reason for this problem is due to the fact that Pod deletion requires
+kubelet's acknowlegement. When a Pod is eviceted (deleted) from a node, the deletion timestamp
+is first set on the Pod. However the Pod will not be actually deleted from API server until kubelet
+finishes cleaning up all the resources of this Pod and kills the running containers. If the node
+is in a failed state and kubelet cannot respond, Pod will be stuck in a teminating state forever.
+In addtion, the containers status for the Pod are still running since kubelet cannot update it,
+which blocks attach_detach controller to detach volumes. The current taint-based eviction actually
+cannot give any benefit in the sense that Pod can only be succesfully evicted after Kubelet comes
+back healthy, in which case there is no need of Pod eviction at all.
+
+
+### Goals
+
+This proposal is trying to investigate the failure handling issues for pods
+and their volumes with a goal of improving system reliability and failure recovery capabilities.
+Among different kinds of failures, node shut down is the main user case this proposal tries
+to address. In this case, we are sure that all containers are stopped, and all mounts are gone
+so that it is safe to take some recovery actions more aggresively without waiting for kubelet correspondence.  
+
+to delete the Pods from API server without waiting for kubelet's acknoledgement.
+Volumes can also be safely detached since mounts no longer exist.
+
+### Non-Goals
+
+This proposal is not trying to address the recovery issue for all types of failures. For example for network
+partition issue, because this is no source of truth about whether Pods are running or not, it is not possible
+for controller decide what recovery actions should take based on the node status.
+
+## Proposal
+
+Currently, Kubernetes already supports four types of taint effects:
+- NoSchedule: Pods that do not tolerate this taint are not scheduled on the node.
+- PreferNoSchedule: Kubernetes avoids scheduling Pods that do not tolerate this taint onto the node.
+- NoExecute: Pod is evicted from the node if it is already running on the node, and is not scheduled
+onto the node if it is not yet running on the node.
+
+We propose to add a new effect, TaintEffectForceEviction. For a Pod that cannot tolerant this taint,
+it will be force deleted from the node with zero grace period which means that Pod can be deleted immediately
+without waiting for Kubelet to responde. One important use case of this taint is node shut down case. If node
+controller or some external controller detact that node is powered off, the controller could add the following
+taint to the node.
+
+
+```
+taint := &v1.Taint{
+	Key:    nodeShutdown,
+	Value:  "true",
+	Effect: v1.TaintEffectForceEviction,
+}
+```
+
+After the node is tainted, node lifecycle controller will trigger to forcefully delete all the Pods from the node.
```

Comment:
/hold

I don’t think this fits with the pod safety design.  I need to think about it, and I’m not comfortable including this without more review.

Pods were supposed to have a toleration for a user defined period of downtime before the node controller force deleted them.  But that *guarantees* you will end up with pod safety issues, especially for statefulsets, and I want to really make sure this is safe